### ISSUE TITLE
httpapi_winhttp.c checking for compiler version instead of OS version

### ISSUE THREAD

**Issue by ewertons** [Azure/azure-c-shared-utility#136]
This code is checking for the version of the compiler, which is incorrect:
https://github.com/Azure/azure-c-shared-utility/commit/edab51e7d3ae6357c635104f7ce14a51e7256a66#diff-50561e196d004a1b8f0c124e2a07f6e0R133

It should be checking the version of the operating system, as described by MSDN:
https://msdn.microsoft.com/en-us/library/windows/desktop/aa384098(v=vs.85).aspx

**Comment by GeraldSchmitt** [comment]
Just in case: there's a partial fix in this PR #132. (removed compiler check, but no os check)

**Comment by PeterSmithson** [comment]
Hi, Just commenting as I submitted PR 132.  If you are not using the solution in PR 132, can I request that you do a runtime check on the OS rather than a compile time check?  Otherwise, we have to ship 2 different binaries depending on what OS the customer has. (Though with a pure runtime check, you will have an issue compiling the code on a Windows 7 PC). The code in PR 132 will work fine when compiled on Windows 10 and running on Windows 7 or Windows 10.  It will also compile OK on Windows 7.  Thanks.

**Comment by ewertons** [comment]
Hi @PeterSmithson ,
I see your point about having a binary that runs on both systems, however the documentation mentions that WINHTTP_ACCESS_TYPE_DEFAULT_PROXY is deprecated on Win8.1 and later. For now the value is still defined in the Win HTTP API, but since it's deprecated in can be removed anytime, so we should not use it. By following that guidance I don't see a way to compile it on Win8.1 and later and make it back compatible using runtime options. 
You mention that the code in PR132 would run OK on Windows 10, however the ifdef check was added exactly because Win10 was not behaving as expected using _DEFAULT_ option (I cannot recall if there were runtime or compilation failures, but I can check).

**Comment by PeterSmithson** [comment]
Hello @ewertons ,
I believe I solved that issue in PR132 by using the new AUTOMATIC option when possible.  I've effectively used a combination of a compile time check and a runtime check.  If it's compiled on Windows 10 and ran on Windows 10, it will only use the new option.  If that same code is ran on Windows 7, it will try the new value which will fail, then it falls back to the old DEFAULT option.  If at some future point they remove that DEFAULT setting in the SDK, the code still compiles as I'm #ifdef'ing on that actual value rather than a particular OS.  I thought it was a good way to do it.

**Comment by ewertons** [comment]
Could you point on the PR where are you #ifdef'ing on the actual value?

The only change related to ifdef I see is the removal of the previous one:
https://github.com/Azure/azure-c-shared-utility/pull/132/files#diff-50561e196d004a1b8f0c124e2a07f6e0L133

**Comment by PeterSmithson** [comment]
Sorry - I got the PR's mixed up.  I'm talking about 135.

**Comment by alexmrtn** [comment]
Hi @ewertons 

I see you are taking a look at the proxy issue. I'm a little bit confused because I thought we decided during last week Skype call to take a look at the PR #132 but I saw some activities on #135.

Thanks @PeterSmithson for your help, but I see your PR is quite similar to the one I previously proposed. Plus, I saw there is some things you didn't take into account:
  - You added a `g_ProxyAutomatic` variable but how do you decide to "activate" it or not ? 
  - You added a preprocessor flag `#ifdef WINHTTP_ACCESS_TYPE_AUTOMATIC_PROXY` so you're still checking at compile time the availability of the automatic proxy ? what if you compile on win10 and run on win7 ?

In #132, I propose to use the `IsWindows8Point1OrGreater()` function from `VersionHelpers.h` to determine whether use `WINHTTP_ACCESS_TYPE_DEFAULT_PROXY` or `WINHTTP_ACCESS_TYPE_AUTOMATIC_PROXY`.  ( `WINHTTP_ACCESS_TYPE_DEFAULT_PROXY` is deprecated from 8.1 https://msdn.microsoft.com/en-us/library/windows/desktop/aa384098(v=vs.85).aspx)

@PeterSmithson, you also add the ability to set manually the proxy using the `setOption`, as I did in #132 but I see some mistakes:
  - You propose to set the proxy on `g_SessionHandle`, this means that all connections (created with `HTTPAPI_CreateConnection` function) will be affected by the proxy option. I do not think that it is the expected behavior. In #132, I propose to attach the proxy to a particular ConnectionHandle (as it is done with certificates). As it is not possible to set it directly on the ConnectionHandle (https://msdn.microsoft.com/en-us/library/windows/desktop/aa384066(v=vs.85).aspx), I set the proxy on each new request in  `HTTPAPI_ExecuteRequest`.
  - Your `CloneOption` function does not copy the strings in the `HTTP_PROXY_OPTIONS` structure but only the pointers.

Regards,
Alexis.

**Comment by PeterSmithson** [comment]
Hi @alexmrtn 

Yes I agree that 132 is very similar and will most probably fix all my problems but I'll reply anyway.

>You added a g_ProxyAutomatic variable but how do you decide to "activate" it or not ?
It's activated (set to 1) by default.  It gets turned off if we fall back to WINHTTP_ACCESS_TYPE_DEFAULT_PROXY .  That happens if we try and use WINHTTP_ACCESS_TYPE_AUTOMATIC_PROXY but we are running on a system that does not support it.  So I'm using the failure of that call as a kind of runtime check - like you used the IsWindows8Point1OrGreater call.  It's a bit hacky as I just noticed what happens rather than using some documentation.

>You added a preprocessor flag #ifdef WINHTTP_ACCESS_TYPE_AUTOMATIC_PROXY so you're still checking at compile time the availability of the automatic proxy ? what if you compile on win10 and run on win7 ?
Yes, I believe both our solutions still need a #ifdef of some sort to allow compilation for Windows 7 where that define is not available.  If my code is compiled on win10 and runs on win 7, the call with WINHTTP_ACCESS_TYPE_AUTOMATIC_PROXY will fail (returns NULL) and the code then tries   WINHTTP_ACCESS_TYPE_DEFAULT_PROXY.

In summary, though my solution is a bit ugly, I think it covers more situations than your solution.  I believe your solution will not compile on a Windows 7 PC. (or with the older SDK or whatever it is that causes the new option to be unavailable).

Regarding the manual proxy setting, my colleague did that work.  I'm sure you're right.  Please ignore PR #135 but consider compilation on Windows 7 in PR #132 (if that's required).

Thanks.

Peter.

**Comment by alexmrtn** [comment]
Thanks for the comment, I put back the `  #if defined _MSC_VER && _MSC_VER < 1600 ` as it was indeed missing in my solution. So this is the new behavior:
  - Compile on windows7: no support for automatic proxy
  - Compile on Windows8.1+: check the os version at runtime 

Maybe you can close #135 if it is no more needed. 

Alexis.

**Comment by PeterSmithson** [comment]
I've closed 135.  How will I know when 132 is available?  At the moment, if we have to update, we apply a patch manually.  It will be great when we don't have to do that!  Thanks.

**Comment by danewalton** [comment]
closing as it looks like this was tackled with #132
